drivers/crc/crc_stm32.c

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2024 Brill Power Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(crc_stm32, CONFIG_CRC_HW_LOG_LEVEL);
+
+#include <errno.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/drivers/crc.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <soc.h>
+#include <stm32_ll_crc.h>
+
+#define DT_DRV_COMPAT st_stm32_crc
+
+struct crc_stm32_cfg {
+	CRC_TypeDef *base;
+	struct stm32_pclken pclken;
+};
+
+struct crc_stm32_data {
+	struct k_mutex mutex;
+	k_timeout_t timeout;
+};
+
+static int crc_lock(const struct device *dev, k_timeout_t timeout)
+{
+	struct crc_stm32_data *data = (struct crc_stm32_data *)dev->data;
+
+	return k_mutex_lock(&data->mutex, timeout);
+}
+
+static int crc_unlock(const struct device *dev)
+{
+	struct crc_stm32_data *data = (struct crc_stm32_data *)dev->data;
+	struct k_mutex crc_mtx = data->mutex;
+	int res = 0;
+
+	/**
+	 * The lock count can be greater than 1 if the same thread acquires
+	 * a lock multiple times, so we have to iterate until lock_count
+	 * is zero
+	 */
+	while (crc_mtx.lock_count != 0) {
+		res = k_mutex_unlock(&data->mutex);
+
+		if (res != 0) {
+			return res;
+		}
+	}
```

Comment:
You don't want to do this, the caller has to make sure to lock/unlock properly.

This also has a race condition if an unlock happens between line 50 & 51 in a higher priority thread/interrupt.